src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/References/ReferenceCleanupService.ReferenceHandlers.cs

```diff
@@ -0,0 +1,185 @@
+                    string itemSpecification = GetAttributeItemSpecification(item);
+
+                    references.Add(new ProjectSystemReferenceInfo(_referenceType, itemSpecification, treatAsUsed == "True"));
+                }
+
+                return references;
+            }
+
+            private string GetAttributeTreatAsUsed(KeyValuePair<string, IImmutableDictionary<string, string>> item)
+            {
+                item.Value.TryGetValue("TreatAsUsed", out string treatAsUsed);
+                treatAsUsed = string.IsNullOrEmpty(treatAsUsed) ? "False" : treatAsUsed;
+
+                return treatAsUsed;
+            }
+
+            private string GetAttributeItemSpecification(KeyValuePair<string, IImmutableDictionary<string, string>> item)
+            {
+                item.Value.TryGetValue(_itemSpecification, out string itemSpecification);
+
+                return itemSpecification;
+            }
+
+            public async Task<bool> UpdateReferenceAsync(ConfiguredProject activeConfiguredProject, ProjectSystemReferenceUpdate referenceUpdate, CancellationToken cancellationToken)
+            {
+                bool wasUpdated = false;
+
+                string projectPath = activeConfiguredProject.UnconfiguredProject.FullPath;
+
+                string newValue = referenceUpdate.Action == ProjectSystemUpdateAction.TreatAsUsed ? "True" : "False";
+
+                await activeConfiguredProject.Services.ProjectLockService.WriteLockAsync(async access =>
+                {
+                    var projectOther = await access.GetProjectAsync(activeConfiguredProject, cancellationToken);
+
+                    try
+                    {
+                        var item = projectOther.AllEvaluatedItems.Where(i =>
+                            string.Compare(i.ItemType, _referenceName, StringComparison.OrdinalIgnoreCase) == 0 &&
+                            i.EvaluatedInclude == referenceUpdate.ReferenceInfo.ItemSpecification).First();
+
+                        if (item != null)
+                        {
+                            await access.CheckoutAsync(projectPath);
+
+                            item.SetMetadataValue("TreatAsUsed", newValue);
+
+                            wasUpdated = true;
+                        }
+                    }
+                    catch
+                    {
+                        wasUpdated = false;
+                    }
+                }, cancellationToken);
+
+                return wasUpdated;
+            }
+        }
+
+        private class ProjectReferenceHandler : ReferenceHandler
+        {
+            internal ProjectReferenceHandler() : base(ProjectSystemReferenceType.Project, ProjectReference.SchemaName, ProjectReference.SchemaName, "Identity")
+            { }
+
+            public override async Task<bool> RemoveReferenceAsync(ConfiguredProject configuredProject, ProjectSystemReferenceInfo reference)
+            {
+                Assumes.Present(configuredProject);
```

Comment:
```suggestion
                Requires.NotNull(configuredProject, nameof(configuredProject));
```

Reply:
when should we use one or the other ?
Assumes for services ? and Requires for values ?

Reply:
`Assumes.Present` is for services. `Requires.*` is for argument validation.